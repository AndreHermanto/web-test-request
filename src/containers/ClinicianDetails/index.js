import React, { Component } from 'react';
import { FormGroup } from 'react-bootstrap';
import { 
  initialState, 
  setFormData,  
  addNewHCP, 
  removeHCP, 
  setHCPForm,
  validateClinicianForm
} from './reducer';
import { 
  PageHeading,
  FormButton
} from './../../components/SharedStyle';
import styled from 'styled-components';
import Input from './../../components/Input';


const AdditionalFormBox = styled.div`
  border-bottom: 1px solid #eee;
  border-top: 1px solid #eee;
  padding-top: 10px;
`;

const HCPButton = styled.label`
  cursor: pointer;
  position: relative;
  padding: 7px 15px;
  border: 1px solid #ccc;
  -webkit-transition: border-color .3s ease,color .3s ease,background-color .3s ease;
  transition: border-color .3s ease,color .3s ease,background-color .3s ease;
  color: #00a6b6;
  margin-top: 10px;
`;

const RemoveLabel = styled.span`
  cursor: pointer;
  position: absolute;
  margin-top: 4px;
  margin-left: 8px;
  font-weight: normal;
  font-size: 11px;
  color: #fff;
  padding: 5px;
  background-color: #5cb85c;
  border-radius: .25em;
  &:hover {
    background:#fff;
    color: black;
    border: 1px solid #5cb85c;
  }
`;

/**
* ClinicianDetials form - UI for input clinician details.
* also has UI for requesting copy to other HCP
*/
class ClinicianDetails extends Component {
  constructor(props) {
    super(props);
    this.handleBack = this.handleBack.bind(this);
    this.handleConfirm = this.handleConfirm.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleAddHCP = this.handleAddHCP.bind(this);
    this.handleHCPChange = this.handleHCPChange.bind(this);
    this.state = initialState(props.route.data);
  }

  handleChange(event) {
    this.setState(setFormData(this.state, event.target))
  }

  handleBack() {
    this.props.route.onChange(this);
    this.props.router.push('/step4');
  }

  handleNext(passValidation) {
    if(!passValidation) return false;
    if(this.props.route.isEdited === true)
    {
      this.props.route.onChange(this);
      this.props.router.push('/summary')
    }
    else {
      this.props.route.onChange(this);
      this.props.router.push('/step6'); 
    }
  }

  handleConfirm() {
    this.setState(validateClinicianForm(this.state), () => {    
      var pass = true;
      for (var field in this.state.validation) {
        if(this.state.validation[field].status === 'error') pass = false;
      }
      this.handleNext(pass); 
    });
  }

  handleAddHCP() {
    this.setState(addNewHCP(this.state, this.state.form.copyToHCP));
  }

  handleRemoveHCP(index) {
    this.setState(removeHCP(this.state, this.state.form.copyToHCP, index));
  }

  handleHCPChange(event, index) {
    this.setState(setHCPForm(this.state, this.state.form.copyToHCP, event.target, index))
  }

  // This renders the validation result after confirm button is clicked.
  validate() {
    return this.state.validated && this.state.validation;
  }
  render() {
    return (
        <div>
          <PageHeading>Step 5: Clinician Details</PageHeading>
          <Input
            field="providerNumber"
            label="Provider Number"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            required
          />
          <Input
            field="medicalSpecialty"
            label="Medical Specialty"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            required
          />
          <Input
            field="firstName"
            label="First Name"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            required
          />
          <Input
            field="lastName"
            label="Last Name"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            required
          />
          <Input
            field="organisation"
            label="Name of organisation, or address of practice"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            required
          />
          <Input
            field="phone"
            label="Phone"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            required
          />
          <Input
            field="email"
            label="Email"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            required
          />
          <Input
            field="fax"
            label="Fax"
            onChange={this.handleChange}
            onValidate={this.validate()}
            formState={this.state.form}
            optional
          />
          {
            this.state.form.copyToHCP.map((form, index) => {
              return <AdditionalFormBox key={index}> 
              <PageHeading>HCP {index + 1} <RemoveLabel onClick={() => this.handleRemoveHCP(index)}>Remove</RemoveLabel></PageHeading>
              <Input
                field="additionalFirstName"
                label="First Name"
                onChange={(e) => this.handleHCPChange(e, index)}
                formState={form}
              />
              <Input
                field="additionalLastName"
                label="Last Name"
                onChange={(e) => this.handleHCPChange(e, index)}
                formState={form}
              />
              <Input
                field="additionalOrganisation"
                label="Name of organisation, or address of practice"
                onChange={(e) => this.handleHCPChange(e, index)}
                formState={form}
              />
              <Input
                field="additionalEmail"
                label="Email"
                onChange={(e) => this.handleHCPChange(e, index)}
                formState={form}
              />
              </AdditionalFormBox>
            })
          }
          <FormGroup>
            <HCPButton onClick={this.handleAddHCP}>
              Request a copy of report sent to another HCP
            </HCPButton> 
          </FormGroup>
          <FormButton 
          bsStyle="warning" 
          onClick={this.handleBack}>
            Back
          </FormButton> 
          <FormButton 
          bsStyle="success" 
          type="submit" 
          onClick={this.handleConfirm}
          >
          Confirm
          </FormButton> 
      </div>
    );
  }
}

export default ClinicianDetails;
