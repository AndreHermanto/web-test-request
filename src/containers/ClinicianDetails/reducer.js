import Immutable from 'immutable';
export function initialState(state) {
  return Object.assign({}, {
      form: Immutable.fromJS({
      fax:'',
      email:''
    })
  });
}

export function setFormData(state = initialState, name, data) {
  return Object.assign({}, state, {
    form: state.form.update(name, value => data)
  });
}

export default { 
  initialState,
  setFormData
};