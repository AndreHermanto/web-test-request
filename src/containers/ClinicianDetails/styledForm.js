import styled, { keyframes } from 'styled-components';

export const LabelOut = keyframes`
  0% {
    left: 1px;
    opacity: 1;
    top: 1.2em;
    font-size: 1em;
    font-weight: 300;
  }
  50% {
    font-size: 1em;
    left: 1em;
    opacity: 0;
    top: 1.2em;
    font-weight: 300;
  }
  50.01% {
    font-size: 0.65em;
    left: 1em;
    opacity: 0;
    top: 0.2em;
    color: #00A6B6;
    font-weight: 400;
  }
  100% {
    font-size: 0.7em;
    opacity: 1;
    left: 1px;
    top: 0.2em;
    color: #00A6B6;
    font-weight: 400;
  }
`;

export const LabelIn = keyframes`
  0% {
    left: 1px;
    opacity: 1;
    top: 1.2em;
    font-size: 1em;
    font-weight: 300;
  }
  50% {
    font-size: 1em;
    left: 1em;
    opacity: 0;
    top: 1.2em;
    font-weight: 300;
  }
  50.01% {
    font-size: 0.65em;
    left: 1em;
    opacity: 0;
    top: 0.2em;
    color: #00A6B6;
    font-weight: 400;
  }
  100% {
    font-size: 0.65em;
    opacity: 1;
    left: 1px;
    top: 0.2em;
    color: #00A6B6;
    font-weight: 400;
  }
`;

export const FormBox = styled.div`
    border-color: #e3e7f1;
    box-shadow: 0 1px 3px #e3e7f1, 0 1px 0 #e3e7f1;
    padding: 19px;
`;

export const InputContainer = styled.div`
  position: relative;
  height: 2.9em;
  border: 0px solid #ECECEC;
  border-bottom-width: 1px;
  label {
    animation-name: ${LabelIn};
    animation-duration: 0.35s;
    animation-direction: reverse;
    animation-fill-mode: forwards;
    animation-timing-function: cubic-bezier(0.77, 0, 0.175, 1);
  }
`;

export const Label = styled.label`
  font-family: "Helvetica Neue", Helvetica, Arial;
  display: block;
  position: absolute;
  margin-top: 2px;
  padding: 1px;
  letter-spacing: 0.012em;
  color: #ACACAC;
  font-size: 1em;
`;


export const InputText = styled.input`
  display: block;
  position: absolute;
  box-sizing: border-box;
  width: 100%;
  border: 0;
  top: 1.2em;
  background: none;
  z-index: 1;
  padding: 1px;
  font-size: 1em;
  font-family: "Helvetica Neue", Helvetica, Arial;
  letter-spacing: 0.012em;
  &:focus {
    outline: 0;
  }

  &:focus + label, &:valid + label {
    animation-name: ${LabelOut};
    animation-duration: 0.35s;
    animation-direction: normal;
    animation-fill-mode: forwards;
    animation-timing-function: cubic-bezier(0.77, 0, 0.175, 1);
  } 
`;

export default { 
  LabelOut, LabelIn, FormBox, InputContainer, Label, InputText
}