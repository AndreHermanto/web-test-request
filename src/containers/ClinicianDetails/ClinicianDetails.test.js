import React from 'react';
import renderer from 'react-test-renderer';
import TestUtils from 'react-addons-test-utils';
import ClinicianDetails from './index';

describe('<ClinicianDetails />', function() {
  var props = {
    route: {
      onChange: jest.fn()
    },
    router:[]
  };
  it('renders correctly', function() {
    const tree = renderer.create(<ClinicianDetails />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('handleChange test case: ', () => {
    const view = TestUtils.renderIntoDocument(<ClinicianDetails />);
    const fax = { "target": { 
      "value": "fax",
      "name": "fax",
      "type": "text"
    }};
    view.handleChange(fax);
    expect(view.state.form.fax).toEqual('fax');  

    const checkbox = { "target": { 
      "value": "on",
      "name": "copy",
      "type": "checkbox",
      "checked":true
    }};
    view.handleChange(checkbox);
    expect(view.state.form.copy).toEqual(true);
  });

  test('handleBack works', () =>  {
    var view = TestUtils.renderIntoDocument(React.createElement(ClinicianDetails, props));                                         
    view.handleBack();
    expect(view.props.router.pop()).toEqual('/step2')
  });
  
  test('handleNext works', () => {
    var view = TestUtils.renderIntoDocument(React.createElement(ClinicianDetails, props));                                         
    view.handleNext(true);
    expect(view.props.router.pop()).toEqual('/step6')
  });
});