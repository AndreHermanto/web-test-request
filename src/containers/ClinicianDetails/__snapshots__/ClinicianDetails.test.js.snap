exports[`<ClinicianDetails /> renders correctly 1`] = `
<form>
  <h3
    className="fAmrmf">
    Step 5: Clinician Details
  </h3>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="formControlsTextarea">
      Provider Number 
    </label>
    <input
      className="form-control"
      id="formControlsTextarea"
      name="providerNumber"
      onChange={[Function]}
      type="text" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="formControlsTextarea">
      Medical Specialty 
    </label>
    <input
      className="form-control"
      id="formControlsTextarea"
      name="medicalSpecialty"
      onChange={[Function]}
      type="text" />
  </div>
  <div
    className="form-group">
    <div
      className="row">
      <div
        className="col-md-6">
        <label
          className="control-label"
          htmlFor="formControlsTextarea">
          First Name 
        </label>
        <input
          className="form-control"
          id="formControlsTextarea"
          name="firstName"
          onChange={[Function]}
          type="text" />
      </div>
      <div
        className="col-md-6">
        <label
          className="control-label"
          htmlFor="formControlsTextarea">
          Last Name 
        </label>
        <input
          className="form-control"
          id="formControlsTextarea"
          name="lastName"
          onChange={[Function]}
          type="text" />
      </div>
    </div>
  </div>
  <div
    className="form-group">
    <div
      className="row">
      <div
        className="col-md-12">
        <label
          className="control-label"
          htmlFor="formControlsTextarea">
          Name of organisation, or address of practice 
        </label>
        <input
          className="form-control"
          id="formControlsTextarea"
          name="address"
          onChange={[Function]}
          type="text" />
      </div>
    </div>
  </div>
  <div
    className="form-group">
    <div
      className="row">
      <div
        className="col-md-4">
        <label
          className="control-label"
          htmlFor="formControlsTextarea">
          Phone 
        </label>
        <input
          className="form-control"
          id="formControlsTextarea"
          name="phone"
          onChange={[Function]}
          type="number" />
      </div>
      <div
        className="col-md-4">
        <label
          className="control-label"
          htmlFor="formControlsTextarea">
          Email 
        </label>
        <input
          className="form-control"
          id="formControlsTextarea"
          name="email"
          onChange={[Function]}
          type="email" />
      </div>
      <div
        className="col-md-4">
        <label
          className="control-label"
          htmlFor="formControlsTextarea">
          Fax(optional)
        </label>
        <input
          className="form-control"
          id="formControlsTextarea"
          name="fax"
          onChange={[Function]}
          type="text" />
      </div>
    </div>
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Request a copy of report sent to another HCP
    </label>
     
    <br />
    <div
      className="react-toggle"
      onClick={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}>
      <div
        className="react-toggle-track">
        <div
          className="react-toggle-track-check">
          <svg
            height="11"
            viewBox="0 0 14 11"
            width="14">
            <title>
              switch-check
            </title>
            <path
              d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
              fill="#fff"
              fillRule="evenodd" />
          </svg>
        </div>
        <div
          className="react-toggle-track-x">
          <svg
            height="10"
            viewBox="0 0 10 10"
            width="10">
            <title>
              switch-x
            </title>
            <path
              d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
              fill="#fff"
              fillRule="evenodd" />
          </svg>
        </div>
      </div>
      <div
        className="react-toggle-thumb" />
      <input
        className="react-toggle-screenreader-only"
        name="copy"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox" />
    </div>
  </div>
  <button
    className="iMtXso btn btn-warning"
    disabled={false}
    onClick={[Function]}
    type="button">
    Back
  </button>
  <button
    className="iMtXso btn btn-success"
    disabled={false}
    onClick={[Function]}
    type="submit">
    Confirm
  </button>
</form>
`;
